### ISSUE TITLE
'BestWeights' callback gives worse C-Index scores / 'EarlyStopping' stops training after 1 epoch

### ISSUE THREAD

**Issue by pasalacquaian** [havakv/pycox#113]
Hi there-- great package, and love all the examples :)

I've been training a 'DeepSurv' model using PyCox, and I've noticed something strange. If I use the 'BestWeights' callback during the fit routine, I get far worse C-index scores than if I have no callback. For example, the mean C-index I get on the validation sets from 5-fold cross validation is 0.51 +- 0.16 when I use 'BestWeights', and 0.72 +- 0.04 when I use no callback. I believe 'BestWeights' will return the weights that gave the lowest partial log likelihood, but I am puzzled as to why the best partial log likelihood should give such unstable and poor c-index results. This sort of behaviour is consistently the case when I repeat training. For reference, I am only training for 15 epochs (for now), with a batch size of 50. Each fold is around 143 training cases and 36 validation cases.

I've also been having some strange things happen when I use the 'EarlyStopping' callback. Although it should have a patience of 10, most of my folds end up getting trained for only 1 epoch. Interestingly, my first fold seems to be trained normally, and I don't see training stop after 1 epoch. There are two warnings that come up along the way:
![image](https://user-images.githubusercontent.com/59702377/146613534-a57d0c53-2a5b-480f-901e-c55375b3acf8.png)
...this is after the first fold is fit
![image](https://user-images.githubusercontent.com/59702377/146613578-704283f4-2964-4d79-887f-1c8b8d1b1ed7.png)
...and this is after 3 epochs of the second fold. After this, all other folds are trained for only 1 epoch.

Any advice or insights greatly appreciated!

**Comment by havakv** [comment]
Hi! Thank you for reporting this. It's not good that we see this kind of behaviour. For your first question (the one about concordance of BestWeights), could you post a full example with data, so I can reproduce your results? It's a bit difficult to understand what's going on without code and data. Are you, for example, seeing the same for the example code [here](https://nbviewer.org/github/havakv/pycox/blob/master/examples/cox-ph.ipynb)?

For your second question, have you modified the EarlyStopping to use concordance maybe? I'm not sure why else it would have numba code here. The numba message is probably not that critical, as it looks like it's only concerned about threading.

The final warning you get is an overflow message, which is in general very serious. As it is the prediction that overflows, it not anything in the implementation of the EarlyStopping callback that causes it. It's hard to reason about what might be the cause without having a reproducible example. However, you might want to check your validation data to see which datapoint causes the overflow, as there might be some outliers or something.

I'm sorry that I can't be of more assistance. The only way I would be really able to help is if you post some reproducible examples.

**Comment by pasalacquaian** [comment]
Hi again,

Unfortunately the data is confidential, and I am unable to share it. It's a relatively common format though: there's a set of patients, and each has values for progression (0 or 1) and progression free survival (number of days). The input features are radiomics features calculated per patient. Your response has been helpful regardless though, and I think things are working now (but I don't know why exactly)! In response to your points:

I tried running the example code, and got an error pretty early on:
![image](https://user-images.githubusercontent.com/59702377/147888523-24375893-c803-4481-bc3c-0b120e60ee0b.png)
Not entirely sure what's causing that, but I haven't tried too hard to figure it out yet. My initial guess was something to do with package versions, and I put aside for the time being.

Haven't modified EarlyStopping in any way, just been using the basic  tt.callbacks.EarlyStopping(). I wondered if maybe it was because I was using a torch optimizer and not a tt optimizer, but that didn't seem to change anything. If it helps, my imports are:
![image](https://user-images.githubusercontent.com/59702377/147888740-491992df-8bca-4a2e-a570-d5ca8f1f11ce.png)
And, my training loop looks like:
![image](https://user-images.githubusercontent.com/59702377/147888973-a0f1175b-5441-4b43-a956-af59b567973e.png)
![image](https://user-images.githubusercontent.com/59702377/147888984-1a5d2626-0268-40d8-9ced-1b2bd6021ee8.png)
I am not sure where the numba would be coming from. Everything prior to the training loop is me reading in and cleaning my dataset using pandas, and creating a KFold splitter using Scikit-Learn. You said the message is probably not critical though, so I'm ok to accept it as a strange quirk and not worry about it.

I did a lot of playing around with the splits for cross validation, and I'm embarrassed to admit I didn't keep track of what I tried when. One change I know I made was from a standard KFold splitter to a stratified one. I'm thinking there was probably something particularly unlucky about the samples per fold (and per batch) when I opened this issue that was causing strange overflows. Even now with things working pretty well, the training curve for one of my folds likes to suddenly spike up into the millions before returning to a more normal loss value. I'm guessing there are some inflammatory samples in there that are causing issues, particularly when there's too many of them in the same batch. As I mentioned though, things seem to be working reasonably now though.

Sorry for not being more helpful on my end, but I appreciated the feedback!

**Comment by havakv** [comment]
I'm glad that you got things working, but I'm sorry that we weren't able to get to the bottom of this.

Your pyarrow error is likely a version problem as you suggest.

The tt.optimizers are essentially just syntactic sugar, so using the regular torch optimizers are perfectly fine.

You code looks fine to me. The only place I can see numba is used is in line 172 where the cindex is computed. I think the warning you get only means that you're might not get the benefits of parallelization for this computation (but this only matters for very large datasets).

**Comment by pasalacquaian** [comment]
Thanks for clarifying each of those points. I'll straighten out my package versions, not worry about torch vs torchtuples, and also not worry too hard about the numba warning.

Have a happy new year!
